### DISCUSSION TITLE
Absolutely Essential Realtime Masking

### DISCUSSION THREAD

**Discussion by adamearle** [opentoonz/opentoonz#6465] [Ideas]
**Feature Request**
Column masking ability to be turned back on through the code that is available and refactored for speed up. Masking is super important to animation and this feature not being available forces users to work at a much slower pace increasing budgets significantly.

**Solution**
The ability to create a **"mask column"** that is either vector or raster. This should provide real-time feedback for artists when the animator if is not using **"Preview mode"** or **"Sub-Camera Preview mode"** in the viewer or combo viewer.

Matte effects would be welcomed if they were in real-time as well. This is difficult because if blurs and other kinds of FX nodes maybe work together to create a matte. This is why a simple Column Mask that's either raster or vector would be the best solution here.

**Comment by age3rcm** [comment]
![1573409592920](https://user-images.githubusercontent.com/36406840/158353968-fed75e86-cbca-40f9-bb6f-07cac69ab937.png)
PPPPPPLLLLLLLEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAAAASSSEEEEE

**Comment by F-Burning** [comment]
Very necessary!

**Comment by adamearle** [comment]
@RodneyBaker has pointed me to this report. I still can't help but think that nodes are the wrong direction for masking to happen in real time. 

If the nodes were to do real-time masking, my suggestion would be to make them the same as existing columns except add a checkbox that says mask. This would signify that the current columns would be the mask version of the current column types that already exist.

1. Vector [Masked]
2. Toonz Raster  [Masked]
3. Raster  [Masked]

Implementation of vector masking could be an opportunity to introduce Natron or Blender roto (masking) nodes path system into OpenToonz. At least there would be some code to look at as a launchpad for getting masking started. Something to think about.

Link:
https://github.com/tahoma2d/tahoma2d/pull/361

**Comment by adamearle** [comment]
Vectors could be something like this https://github.com/opentoonz/opentoonz/issues/4198 and https://github.com/opentoonz/opentoonz/issues/4107

![147369752-990c57fd-2fde-479d-a648-8014ec652d9e](https://user-images.githubusercontent.com/16330340/158388023-4ef66522-a321-4aff-8e52-0dd23debb40c.gif)

**Comment by adamearle** [comment]
This could also extend into SVG code https://github.com/opentoonz/opentoonz_docs/issues/199

**Comment by adamearle** [comment]
> now what a mask is but what functionality are you wishing for here? 

yep seems that it was because of this https://github.com/tahoma2d/tahoma2d/pull/361.

Curious to know how you interpret "Real-time masking playback not using Preview mode" or "Sub-Camera Preview mode" in the viewer or combo viewer"?

I am wondering if you think it can be added a node?

**Comment by adamearle** [comment]
Perhaps columns that contain 1 of the 3 different types of data types: Raster, Toonz Raster, and Toonz Vector could have a mask input or check box. This would tell the column node, duplicate, cloned or not to be a mask/ matte, so whatever column connected to it would be a mask. 

The key thing here is that it can playback in real-time.

**Comment by gab3d** [comment]
Ultimately the goal should be to have realtime (or at least 'super fast') previewing in OT, so that by default the Preview mode could be ON.

**Comment by F-Burning** [comment]
The problem I see with columns is the space in the Xsheet/timeline, because create columns every time you need a mask maybe could be chaotic. As @gab3d said, the point would be works with nodes in real time instead of previews, but it is such an important change that I think it will never be done, unfortunately.

I have realized that Opentoonz has such a specific way of working, that you can't do anything that goes out of line because it will be problematic.

**Comment by adamearle** [comment]
> 

That's incredibly disheartening. Does this mean we shouldn't try?

As for columns stacking. I am sure users will have no problem with that Adobe After Effects does exactly the same thing, I know that I won't. I would rather have something that I can work within real-time than nothing at all.

I do understand what you are saying. Though in order to attract more development we really need to be getting something down. I agree with you nodes would be great. If that is the fastest and easiest direction to go then I would say,

                                              LETS DOOOOO IIIIIT!
                                                                                
If it's not then I honestly don't really care. The idea is masking in real-time. If means the matte in and out nodes are rewritten then kool. I am sure I don't need to tell you OT needs this.

Fixing mask, Fixing keys is going to make life easier. Which means more users. Which means more development. Which means more money. Which means a better tool. I am trusting that the developer will be able to do what they do best and spot the problems. So we can all start working out how to make it happen.

Even if it means passing the hat around getting someone to work on it for 3 to 4 months so it's solid. It has to start somewhere.

**Comment by shun-iwasawa** [comment]
I'm planning to develop the mask feature in the near future (probably will work on a prototype in V1.7 period).

My blueprint of the feature would be similar to the one in Adobe AfterEffects.
That is to say, the mask shape will be specified by bezier curve (i.e. a vector shape with no line thickness). And it will be more a "property" of the column rather than the "mask column"; a column with special purpose.

Yes, I would like to make the mask to take effect in real-time which will be handy, but I think the major advantage of this feature over the Matte Fx would be that users do not need to save a level file for mask shape since the mask shape (and its animation) will be saved in the scene file.

**Comment by gab3d** [comment]
@shun-iwasawa 
great to know! 
will there be some sort of "feather" and "expansion" features for those masks, as in AE?

**Comment by RodneyBaker** [comment]
Tagging with roadmap as @shun-iwasawa has stated plans to work on related area.
As we get more information on that effort we can assess how closely it addresses the requirements of AERM (TM) as well as that  requested by @konero, @melieconiek and others.

**Comment by adamearle** [comment]
That is an interesting idea as well on the mask @shun-iwasawa. I might be worthwhile  building an option in there to choose:

1. Bezier mask 
2.Choose Colum

Choose Column may come with a warning, "Playback may be slow because of a complex connection made in the schematic view".

You may also want to take a look at this project for some ideas? https://github.com/MaurycyLiebner/enve I spent yesterday afternoon looking at this and pushing it. The only thing that i would say about the "enve" Project is that there can't be new bezier objects from frame to frame. I have outlined that here https://github.com/MaurycyLiebner/enve/issues/302

There are some nice ideas in this project for you to build upon and play with. I hope you like it.

**Comment by konero** [comment]
I only just stumbled on this thread. I don't know how well you guys know Toon Boom Harmony, but in this software they handle this with a node (sort of), it's called the **Animated Matte Generator** and just like Toonz you can filter out specific color indexes you want to attach the animated matte to.

There's a neat video of a user explaining how to use it in detail here: https://youtu.be/g_PyeLb4czg?t=225

**Comment by adamearle** [comment]
@konero yeah I have been a TBH rigger from the way. Making with nodes makes a lot of sense to me. many applications take this approach as well. I think a real-time masking node could be implemented.

**Comment by RodneyBaker** [comment]
This is a note to let everyone know this general feature of column masking is present in a working release of Tahoma2D.
As some aspects of the feature will need additional work (such as how column masking works with FX) inclusion is not expected for some time in a final release of T2D. 

For reference please see:  https://github.com/tahoma2d/tahoma2d/pull/1142

**Comment by RodneyBaker** [comment]
In addition to vector level masks @manongjohn has added smart raster (toonz raster) and raster levels as column masks.
There is still some work to be done so the Pull Request is WIP for now with target release of T2D v1.5 (Dec 2024 timeframe).

REF:  https://github.com/tahoma2d/tahoma2d/pull/1307 

The vector level real time column mask can be found in current T2D nightly releases
